File: web/src/lib/clima.ts
/**
 * Clima diario e o balanco de agua no solo: a parte pura, sem rede.
 *
 * A busca no Open-Meteo mora em `open-meteo.ts`. Aqui ficam as contas, porque
 * sao elas que precisam de teste e a suite deste projeto nao toca rede.
 *
 * O QUE O MODELO NOVO EXIGIU DAQUI
 * --------------------------------
 * O modelo antigo pedia sete medias de uma janela. O `modelo_gramas.pkl` pede
 * `agua_solo_media_pct` e `dias_encharcado`, que sao ESTADO: nao saem de uma
 * media, saem de um balde de agua rodando dia a dia. E balde precisa de
 * AQUECIMENTO -- comecar hoje e chutar quanta agua ha no solo agora, e essa
 * feature move o crescimento previsto em mais de 100% entre 10% e 55% de balde
 * cheio. Por isso `Janela` passou a carregar os dias ANTERIORES a hoje.
 *
 * A distincao entre TOTAL e MEDIA continua sendo a conta que mais importa. O
 * modelo recebe `precipitacao_total_mm` (soma do periodo) junto com
 * `dias_periodo`. Alimentar o total de uma janela de 16 dias declarando 90 dias
 * faz o modelo enxergar uma seca de dois meses e prever crescimento perto de
 * zero, com a mesma cara de certeza de sempre. E por isso que a janela tem que
 * ter o tamanho do periodo simulado.
 */

/** Teto da API de previsao do Open-Meteo. */
export const DIAS_DE_PREVISAO = 16;

/**
 * Quantos dias de passado pedir para aquecer o balde.
 *
 * A API de previsao aceita 92 e entrega ~63 uteis (antes disso ela devolve
 * nulo, e `lerDiario` descarta). Sessenta e tres bastam: com ET0 de ~3 mm/dia
 * um balde de 60 mm se esvazia em vinte. O notebook de calibracao usa 120 e vai
 * ao arquivo ERA5 para conseguir; aqui uma requisicao a menos vale mais que os
 * 57 dias extras, porque o arquivo e justamente o que responde 429.
 */
export const DIAS_DE_AQUECIMENTO = 92;

export type FonteDia =
  /** Dia que ja passou, com o dado que a propria API de previsao guarda. */
  | "observado"
  /** Previsao de verdade para aquele dia. */
  | "previsao"
  /** Media do mesmo dia do calendario em anos anteriores (ERA5 observado). */
  | "historico"
  /** O padrao dos dias previstos, repetido, a queda quando o historico falha. */
  | "repeticao";

export type DiaClima = {
  /** AAAA-MM-DD. */
  data: string;
  temperaturaC: number;
  temperaturaMinC: number;
  temperaturaMaxC: number;
  umidadePct: number;
  chuvaMm: number;
  radiacaoMjM2: number;
  et0MmDia: number;
  fonte: FonteDia;
};

/** O formato bruto que as duas APIs do Open-Meteo devolvem, a de previsao e a
 *  de arquivo usam exatamente os mesmos nomes de variavel. */
export type RespostaDiaria = {
  daily?: {
    time?: string[];
    temperature_2m_mean?: (number | null)[];
    temperature_2m_min?: (number | null)[];
    temperature_2m_max?: (number | null)[];
    relative_humidity_2m_mean?: (number | null)[];
    precipitation_sum?: (number | null)[];
    shortwave_radiation_sum?: (number | null)[];
    et0_fao_evapotranspiration?: (number | null)[];
  };
  error?: boolean;
  reason?: string;
};

/**
 * Converte a resposta em dias utilizaveis.
 *
 * Dia com qualquer variavel nula e DESCARTADO, em vez de virar zero. Zero nao e
 * "nao sei": zero de radiacao e noite polar, zero de et0 e ar saturado, e os
 * dois puxariam a media para baixo fingindo medicao. E tambem o que separa o
 * passado que a API tem do passado que ela nao tem: pedir 92 dias e receber 63
 * com dado e 29 nulos e o caso normal, nao erro.
 */
export function lerDiario(resposta: RespostaDiaria, fonte: FonteDia): DiaClima[] {
  const d = resposta.daily;
  if (!d?.time) return [];

  const dias: DiaClima[] = [];

  for (let i = 0; i < d.time.length; i += 1) {
    const valores = [
      d.temperature_2m_mean?.[i],
      d.temperature_2m_min?.[i],
      d.temperature_2m_max?.[i],
      d.relative_humidity_2m_mean?.[i],
      d.precipitation_sum?.[i],
      d.shortwave_radiation_sum?.[i],
      d.et0_fao_evapotranspiration?.[i],
    ];
    if (valores.some((v) => v == null || !Number.isFinite(v))) continue;

    dias.push({
      data: d.time[i],
      temperaturaC: valores[0] as number,
      temperaturaMinC: valores[1] as number,
      temperaturaMaxC: valores[2] as number,
      umidadePct: valores[3] as number,
      chuvaMm: valores[4] as number,
      radiacaoMjM2: valores[5] as number,
      et0MmDia: valores[6] as number,
      fonte,
    });
  }

  return dias;
}

function media(valores: number[]): number {
  return valores.length ? valores.reduce((a, b) => a + b, 0) / valores.length : 0;
}

/* ------------------------------------------------------------------ *
 * Balanco de agua no solo
 * ------------------------------------------------------------------ */

export type Balanco = {
  /** Quanto o balde esta cheio em cada dia, de 0 a 1. */
  fracoes: number[];
  /** Tres dias seguidos de balde quase cheio com chuva forte. */
  encharcado: boolean[];
};

/**
 * Balde FAO simplificado, dia a dia.
 *
 * Porte fiel do `balanco_solo` de `clima.py`, que por sua vez repete o do
 * `gerador_v3_1_rebrota.py` -- as tres copias precisam continuar iguais, ou
 * `agua_solo_media_pct` passa a significar coisas diferentes no treino, no lote
 * e no simulador.
 *
 * Uma diferenca deliberada em relacao ao gerador: o Kc fica FIXO na altura
 * inicial em vez de acompanhar a altura dia a dia. No gerador a altura e
 * conhecida em todo dia porque ele a simula; na inferencia ela e justamente o
 * que se quer prever, entao fixar no ponto de partida e o que da para fazer sem
 * circularidade.
 */
export function balancoSolo(
  dias: readonly DiaClima[],
  capacidadeMm: number,
  alturaCm: number,
): Balanco {
  const cap = Math.max(capacidadeMm, 1e-6);
  let agua = cap * 0.6; // comeco neutro, como no gerador
  const kc = 0.4 + 0.6 * Math.min(Math.max(alturaCm, 0) / 40, 1);

  const fracoes: number[] = [];
  const encharcado: boolean[] = [];
  let seguidos = 0;

  for (const d of dias) {
    agua = Math.min(agua + d.chuvaMm, cap);
    const ks = Math.min(Math.max(agua / (0.55 * cap), 0), 1);
    agua = Math.min(Math.max(agua - d.et0MmDia * kc * ks, 0), cap);
    const f = agua / cap;
    seguidos = f > 0.97 && d.chuvaMm > 8 ? seguidos + 1 : 0;
    fracoes.push(f);
    encharcado.push(seguidos >= 3);
  }

  return { fracoes, encharcado };
}

/* ------------------------------------------------------------------ *
 * Agregados de tela
 * ------------------------------------------------------------------ */

/** O resumo que a tela mostra da janela. As features do modelo sao outra coisa
 *  e saem de `montarFeatures`, em `modelo/campos.ts`. */
export type ResumoClima = {
  dias: number;
  temperaturaMediaC: number;
  temperaturaMinC: number;
  temperaturaMaxC: number;
  umidadeMediaPct: number;
  precipitacaoTotalMm: number;
  diasComChuva: number;
  radiacaoMediaMjM2: number;
  et0MedioMmDia: number;
  /** Media do balde cheio no periodo, de 0 a 100. */
  aguaSoloMediaPct: number | null;
};

export function resumir(dias: readonly DiaClima[], fracoes?: readonly number[]): ResumoClima {
  const n = dias.length;
  if (n === 0) {
    throw new Error("Não dá para resumir uma janela de clima vazia.");
  }

  return {
    dias: n,
    temperaturaMediaC: media(dias.map((d) => d.temperaturaC)),
    temperaturaMinC: Math.min(...dias.map((d) => d.temperaturaMinC)),
    temperaturaMaxC: Math.max(...dias.map((d) => d.temperaturaMaxC)),
    umidadeMediaPct: media(dias.map((d) => d.umidadePct)),
    precipitacaoTotalMm: dias.reduce((total, d) => total + d.chuvaMm, 0),
    diasComChuva: dias.filter((d) => d.chuvaMm > 1).length,
    radiacaoMediaMjM2: media(dias.map((d) => d.radiacaoMjM2)),
    et0MedioMmDia: media(dias.map((d) => d.et0MmDia)),
    aguaSoloMediaPct: fracoes?.length ? media([...fracoes]) * 100 : null,
  };
}

/* ------------------------------------------------------------------ *
 * Montagem da janela
 * ------------------------------------------------------------------ */

/**
 * Media dia a dia entre varios anos, alinhada por POSICAO e nao por data.
 *
 * Por posicao porque 29 de fevereiro desalinha o calendario: a janela de um ano
 * bissexto tem um dia a mais que a do ano seguinte, e casar por data deixaria
 * buracos no meio da serie. O que interessa aqui e "o enesimo dia depois do
 * inicio", que e o que a simulacao consome.
 *
 * As datas devolvidas sao as do PRIMEIRO ano da lista, so para a serie ter
 * rotulo; quem chama sobrescreve com as datas futuras de verdade.
 */
export function mediaEntreAnos(anos: DiaClima[][]): DiaClima[] {
  const uteis = anos.filter((a) => a.length > 0);
  if (uteis.length === 0) return [];

  const comprimento = Math.min(...uteis.map((a) => a.length));
  const saida: DiaClima[] = [];

  for (let i = 0; i < comprimento; i += 1) {
    const doDia = uteis.map((a) => a[i]);
    saida.push({
      data: uteis[0][i].data,
      temperaturaC: media(doDia.map((d) => d.temperaturaC)),
      // Minima e maxima do periodo sao EXTREMOS no modelo, e media de extremos
      // entre anos ja e o que se quer: um dia frio isolado em 2024 nao deve
      // definir sozinho a minima de uma janela de 2026.
      temperaturaMinC: media(doDia.map((d) => d.temperaturaMinC)),
      temperaturaMaxC: media(doDia.map((d) => d.temperaturaMaxC)),
      umidadePct: media(doDia.map((d) => d.umidadePct)),
      chuvaMm: media(doDia.map((d) => d.chuvaMm)),
      radiacaoMjM2: media(doDia.map((d) => d.radiacaoMjM2)),
      et0MmDia: media(doDia.map((d) => d.et0MmDia)),
      fonte: "historico",
    });
  }

  return saida;
}

/**
 * Estica a previsao repetindo-a ciclicamente ate `total` dias.
 *
 * Ciclicamente, e nao com a media dos 16 dias em todos os dias, porque o modelo
 * le a variabilidade: uma serie chapada na media tem a mesma temperatura media
 * mas nenhum dia de chuva forte, e `dias_com_chuva` e o balde mudam. Repetir o
 * ciclo preserva media e dispersao.
 *
 * E a queda, nao a primeira escolha: o historico observado e melhor. Mas ela
 * nunca falha, e numa tela de demonstracao um 429 do Open-Meteo nao pode virar
 * pagina de erro.
 */
export function completarPorRepeticao(base: DiaClima[], total: number, datas: string[]): DiaClima[] {
  if (base.length === 0) return [];

  const saida = [...base];
  for (let i = base.length; i < total; i += 1) {
    const molde = base[i % base.length];
    saida.push({ ...molde, data: datas[i] ?? molde.data, fonte: "repeticao" });
  }
  return saida.slice(0, total);
}

export type Janela = {
  /** Dias ANTERIORES a hoje, so para aquecer o balde de agua no solo. Nao
   *  entram em nenhuma media da tela nem em nenhuma feature de periodo. */
  aquecimento: DiaClima[];
  /** Os dias simulados, a partir de hoje. */
  dias: DiaClima[];
  /** Quantos dias vieram de previsao de verdade. */
  diasPrevistos: number;
  /** Como o resto foi preenchido. `null` quando a previsao cobriu tudo. */
  complemento: "historico" | "repeticao" | null;
  /** Anos do ERA5 que entraram na media, quando o complemento foi historico. */
  anos: number[];
  /** Por que caiu para a repeticao, quando caiu. Vai para a tela. */
  avisoDoComplemento: string | null;
};

/**
 * Monta a janela final de `total` dias: previsao na frente, complemento atras.
 *
 * As datas do complemento sao sempre as datas FUTURAS de verdade, o historico
 * empresta os numeros, nao o calendario.
 */
export function montarJanela(entrada: {
  aquecimento?: DiaClima[];
  previsao: DiaClima[];
  historico: DiaClima[];
  anos: number[];
  total: number;
  datas: string[];
  avisoDoComplemento?: string | null;
}): Janela {
  const { previsao, historico, anos, total, datas } = entrada;
  const aquecimento = entrada.aquecimento ?? [];

  if (previsao.length === 0) {
    throw new Error("A previsão do Open-Meteo voltou sem nenhum dia utilizável.");
  }

  const previstos = Math.min(previsao.length, total);
  const cabeca = previsao.slice(0, previstos);

  if (previstos >= total) {
    return {
      aquecimento,
      dias: cabeca,
      diasPrevistos: previstos,
      complemento: null,
      anos: [],
      avisoDoComplemento: null,
    };
  }

  const faltam = total - previstos;
  const cauda = historico.slice(0, faltam).map((d, i) => ({
    ...d,
    data: datas[previstos + i] ?? d.data,
  }));

  if (cauda.length === faltam) {
    return {
      aquecimento,
      dias: [...cabeca, ...cauda],
      diasPrevistos: previstos,
      complemento: "historico",
      anos,
      avisoDoComplemento: entrada.avisoDoComplemento ?? null,
    };
  }

  return {
    aquecimento,
    dias: completarPorRepeticao(cabeca, total, datas),
    diasPrevistos: previstos,
    complemento: "repeticao",
    anos: [],
    avisoDoComplemento:
      entrada.avisoDoComplemento ??
      "O arquivo histórico do Open-Meteo não respondeu a tempo.",
  };
}

File: clima.py
"""
O preenchedor: de duas chamadas ao Open-Meteo para as 17 features de clima.

O modelo antigo pedia sete agregados de uma janela de 16 dias, e `buscar_clima`
cabia em vinte linhas. O `modelo_gramas.pkl` mudou isso de figura: ele pede
`agua_solo_media_pct` e `dias_encharcado`, que sao ESTADO -- nao dao para
calcular a partir de uma media, so a partir de uma serie DIARIA rodando um
balanco de agua no solo. Este arquivo e a porta do preenchedor do notebook de
calibracao (`docs/previsao_gramas_colab.ipynb`, celula 3) para dentro do lote.

TRES PEDACOS DE SERIE, TRES ORIGENS
-----------------------------------
    [hoje-63, hoje-1]   passado real, da propria API de previsao (`past_days`)
    [hoje,  hoje+15]    previsao de verdade
    [hoje+16, hoje+120] ERA5 do ano passado, mesmos dias do calendario

O passado nao e enfeite: o balde de agua no solo precisa de AQUECIMENTO. Comecar
o balde no dia de hoje e chutar quanta agua ha no solo, e `agua_solo_media_pct`
move o q50 em mais de 100% entre 10% e 55%. O notebook usa 120 dias de
aquecimento; a API de previsao entrega ~63 (pedimos 92 e ela devolve nulo antes
disso), e 63 bastam porque o balde tem constante de tempo de semanas: com ET0 de
~3 mm/dia um balde de 60 mm se esvazia em 20 dias.

`past_days=92` numa chamada so, em vez de uma chamada ao arquivo ERA5 para o
aquecimento: e uma requisicao a menos por zona, e o arquivo e justamente o que
responde 429 quando se insiste (ver o cabecalho de `web/src/lib/open-meteo.ts`).

A CAUDA LONGA E POR CAUSA DO `dias_ate_limite`
----------------------------------------------
O alvo do modelo e `crescimento_total_cm` de um periodo, e a resposta e nao
linear no tamanho do periodo. Para saber em que dia o trecho cruza a altura
limite nao da mais para dividir por um cm/dia: e preciso perguntar ao modelo
"quanto cresce em 1 dia? em 2? em 3?" ate cruzar. Isso exige clima ate 120 dias
a frente, e previsao so vai a 16. Dai o ERA5 do ano passado nos mesmos dias do
calendario -- que erra o tempo daquela semana e acerta a ESTACAO, que e o que
importa num horizonte de tres meses. Quando ele recusa, o padrao dos 16 dias
previstos se repete ciclicamente, e a serie diz qual dos dois valeu.
"""

import time
from datetime import date, timedelta
from typing import Iterable, NamedTuple

import httpx

API_PREVISAO = "https://api.open-meteo.com/v1/forecast"
API_ARQUIVO = "https://archive-api.open-meteo.com/v1/archive"

DIARIAS = ("temperature_2m_mean,temperature_2m_min,temperature_2m_max,"
           "relative_humidity_2m_mean,precipitation_sum,"
           "shortwave_radiation_sum,et0_fao_evapotranspiration")

#: Quanto passado pedir a API de previsao. Ela aceita 92 e entrega ~63 uteis.
AQUECIMENTO_DIAS = 92
#: Teto da API de previsao.
PREVISAO_DIAS = 16
#: Horizonte da busca do cruzamento. E o teto de `dias_periodo` no treino.
HORIZONTE_DIAS = 120

TIMEOUT_S = 45.0
PAUSA_S = 0.6

# Parametros por especie que entram nas CONTAS das features (nao no modelo).
# Copiados de `gerador_v3_1_rebrota.py`. `t_base` e `t_ot2` definem os graus-dia
# e `geada` o limiar de Tmin que conta como geada -- os tres precisam bater com
# o gerador, ou a feature muda de significado entre treino e producao.
ESPECIES = {
    "braquiaria": dict(t_base=15.0, t_ot2=35.0, geada=2.0,
                       flor_meses=(2, 3, 4), flor_hmin=30.0),
    "esmeralda": dict(t_base=12.0, t_ot2=32.0, geada=-2.0,
                      flor_meses=(), flor_hmin=99.0),
    "batatais": dict(t_base=13.0, t_ot2=33.0, geada=0.0,
                     flor_meses=(11, 12, 1, 2, 3), flor_hmin=10.0),
}


class Dia(NamedTuple):
    data: date
    tmed: float
    tmin: float
    tmax: float
    umidade: float
    chuva: float
    radiacao: float
    et0: float
    #: "observado" | "previsao" | "historico" | "repeticao"
    fonte: str


class Serie(NamedTuple):
    dias: list[Dia]
    #: Quantos dias de aquecimento (anteriores a hoje) entraram.
    aquecimento: int
    #: Como a cauda alem da previsao foi preenchida. `None` = nao precisou.
    complemento: str | None
    #: Ano do ERA5 que emprestou a cauda, quando o complemento foi historico.
    ano_historico: int | None
    #: Por que caiu para a repeticao, quando caiu.
    aviso: str | None

    def indice_de(self, d: date) -> int | None:
        for i, dia in enumerate(self.dias):
            if dia.data == d:
                return i
        return None


# ----------------------------------------------------------------------
# Busca
# ----------------------------------------------------------------------
def _pedir(url: str, params: dict) -> dict:
    r = httpx.get(url, params=params, timeout=TIMEOUT_S)
    r.raise_for_status()
    corpo = r.json()
    if corpo.get("error"):
        raise RuntimeError(corpo.get("reason", "Open-Meteo recusou a consulta."))
    return corpo


def _ler(corpo: dict, fonte: str) -> list[Dia]:
    """Dia com QUALQUER variavel nula e descartado, nao zerado.

    Zero nao e "nao sei": zero de radiacao e noite polar, zero de et0 e ar
    saturado, e os dois puxariam a media para baixo fingindo medicao. E o mesmo
    criterio de `lerDiario` no painel.
    """
    d = corpo.get("daily") or {}
    tempos = d.get("time") or []
    saida: list[Dia] = []

    for i, t in enumerate(tempos):
        v = [
            (d.get("temperature_2m_mean") or [None])[i] if i < len(d.get("temperature_2m_mean") or []) else None,
            (d.get("temperature_2m_min") or [None])[i] if i < len(d.get("temperature_2m_min") or []) else None,
            (d.get("temperature_2m_max") or [None])[i] if i < len(d.get("temperature_2m_max") or []) else None,
            (d.get("relative_humidity_2m_mean") or [None])[i] if i < len(d.get("relative_humidity_2m_mean") or []) else None,
            (d.get("precipitation_sum") or [None])[i] if i < len(d.get("precipitation_sum") or []) else None,
            (d.get("shortwave_radiation_sum") or [None])[i] if i < len(d.get("shortwave_radiation_sum") or []) else None,
            (d.get("et0_fao_evapotranspiration") or [None])[i] if i < len(d.get("et0_fao_evapotranspiration") or []) else None,
        ]
        if any(x is None for x in v):
            continue
        saida.append(Dia(date.fromisoformat(t), float(v[0]), float(v[1]), float(v[2]),
                         float(v[3]), float(v[4]), float(v[5]), float(v[6]), fonte))
    return saida


def _menos_um_ano(d: date) -> date:
    try:
        return d.replace(year=d.year - 1)
    except ValueError:          # 29 de fevereiro
        return d.replace(year=d.year - 1, month=2, day=28)


def buscar_serie(lat: float, lon: float, hoje: date,
                 horizonte: int = HORIZONTE_DIAS) -> Serie:
    """Aquecimento + previsao + cauda, ja emendados e ordenados por data."""
    corpo = _pedir(API_PREVISAO, {
        "latitude": lat, "longitude": lon, "daily": DIARIAS,
        "past_days": AQUECIMENTO_DIAS, "forecast_days": PREVISAO_DIAS,
        "timezone": "America/Sao_Paulo",
    })
    bruto = _ler(corpo, "previsao")
    dias = [d._replace(fonte="observado") if d.data < hoje else d for d in bruto]
    dias = [d for d in dias if d.data <= hoje + timedelta(days=horizonte - 1)]
    if not dias:
        raise RuntimeError("A previsao do Open-Meteo voltou sem nenhum dia utilizavel.")

    aquecimento = sum(1 for d in dias if d.data < hoje)
    ultimo = max(d.data for d in dias)
    alvo = hoje + timedelta(days=horizonte - 1)

    if ultimo >= alvo:
        return Serie(dias, aquecimento, None, None, None)

    faltam_de = ultimo + timedelta(days=1)
    complemento, ano, aviso = "historico", None, None
    cauda: list[Dia] = []

    try:
        time.sleep(PAUSA_S)
        corpo = _pedir(API_ARQUIVO, {
            "latitude": lat, "longitude": lon, "daily": DIARIAS,
            # Cinco dias de sobra: ano bissexto muda o comprimento da janela em
            # um dia, e faltar um dia derrubaria a cauda inteira na repeticao.
            "start_date": _menos_um_ano(faltam_de).isoformat(),
            "end_date": _menos_um_ano(alvo + timedelta(days=5)).isoformat(),
            "timezone": "America/Sao_Paulo",
        })
        passado = _ler(corpo, "historico")
        precisa = (alvo - faltam_de).days + 1
        if len(passado) >= precisa:
            ano = passado[0].data.year
            # O historico empresta os NUMEROS, nao o calendario.
            cauda = [d._replace(data=faltam_de + timedelta(days=i))
                     for i, d in enumerate(passado[:precisa])]
        else:
            aviso = "O arquivo historico do Open-Meteo devolveu menos dias que o periodo."
    except Exception as e:
        aviso = f"O arquivo historico do Open-Meteo nao respondeu ({type(e).__name__})."

    if not cauda:
        complemento = "repeticao"
        # Ciclicamente, e nao com a media dos 16 dias em todos: uma serie
        # chapada na media tem a mesma temperatura media e nenhum dia de chuva
        # forte, e `dias_com_chuva` e o balde mudam. Repetir preserva os dois.
        futuros = [d for d in dias if d.data >= hoje] or dias[-PREVISAO_DIAS:]
        n = (alvo - faltam_de).days + 1
        cauda = [futuros[i % len(futuros)]._replace(
            data=faltam_de + timedelta(days=i), fonte="repeticao") for i in range(n)]

    return Serie(dias + cauda, aquecimento, complemento, ano, aviso)


# ----------------------------------------------------------------------
# Balanco de agua no solo
# ----------------------------------------------------------------------
def balanco_solo(dias: Iterable[Dia], capacidade_mm: float,
                 altura_cm: float) -> tuple[list[float], list[bool]]:
    """Balde FAO simplificado, dia a dia. Devolve (fracao cheia, encharcado).

    Identico ao `balanco_solo` do notebook, que por sua vez repete o do
    `gerador_v3_1_rebrota.py` com uma diferenca deliberada: o Kc fica FIXO na
    altura inicial em vez de acompanhar a altura dia a dia. No gerador a altura
    e conhecida em todo dia porque ele a simula; na inferencia ela e justamente
    o que se quer prever, entao fixar no ponto de partida e o que da para fazer
    sem circularidade.
    """
    cap = max(float(capacidade_mm), 1e-6)
    agua = cap * 0.6                      # comeco neutro, como no gerador
    kc = 0.4 + 0.6 * min(max(altura_cm, 0.0) / 40.0, 1.0)
    fracoes: list[float] = []
    encharcado: list[bool] = []
    seguidos = 0

    for d in dias:
        agua = min(agua + d.chuva, cap)
        ks = min(max(agua / (0.55 * cap), 0.0), 1.0)
        agua = min(max(agua - d.et0 * kc * ks, 0.0), cap)
        f = agua / cap
        seguidos = seguidos + 1 if (f > 0.97 and d.chuva > 8.0) else 0
        fracoes.append(f)
        encharcado.append(seguidos >= 3)

    return fracoes, encharcado


# ----------------------------------------------------------------------
# Montagem das features
# ----------------------------------------------------------------------
def montar_features(*, especie: str, altura_cm: float, dias_desde_rocada: float,
                    latitude: float, serie: Serie, inicio: date, dias_periodo: int,
                    fertilidade: float, capacidade_mm: float,
                    fracoes: list[float] | None = None,
                    encharcado: list[bool] | None = None) -> dict:
    """As 20 features, para a janela [inicio, inicio + dias_periodo).

    `fracoes`/`encharcado` sao o balanco de solo JA calculado sobre a serie
    inteira. Vem de fora porque o balanco depende da serie e da altura inicial,
    e nao do tamanho da janela: quem varre 120 horizontes calcula uma vez so.
    """
    if dias_periodo < 1:
        raise ValueError("dias_periodo precisa ser pelo menos 1.")

    e = ESPECIES[especie]
    fim = inicio + timedelta(days=dias_periodo)
    janela = [(i, d) for i, d in enumerate(serie.dias) if inicio <= d.data < fim]
    if not janela:
        raise RuntimeError(f"A serie de clima nao cobre [{inicio}, {fim}).")

    if fracoes is None or encharcado is None:
        fracoes, encharcado = balanco_solo(serie.dias, capacidade_mm, altura_cm)

    idx = [i for i, _ in janela]
    dd = [d for _, d in janela]
    n = len(dd)

    graus_dia = sum(max(min(d.tmed, e["t_ot2"]) - e["t_base"], 0.0) for d in dd)
    floracao = sum(1 for d in dd
                   if d.data.month in e["flor_meses"] and altura_cm > e["flor_hmin"])

    return {
        "especie": especie,
        "dias_periodo": n,
        "altura_inicial_cm": float(altura_cm),
        "dias_desde_rocada_inicio": float(dias_desde_rocada),
        "temperatura_media_c": round(sum(d.tmed for d in dd) / n, 1),
        "temperatura_min_c": round(min(d.tmin for d in dd), 1),
        "temperatura_max_c": round(max(d.tmax for d in dd), 1),
        "graus_dia_acumulados": round(graus_dia, 1),
        "umidade_media_pct": round(sum(d.umidade for d in dd) / n, 1),
        "precipitacao_total_mm": round(sum(d.chuva for d in dd), 1),
        "dias_com_chuva": sum(1 for d in dd if d.chuva > 1.0),
        "et0_medio_mm_dia": round(sum(d.et0 for d in dd) / n, 2),
        "radiacao_media_mj_m2": round(sum(d.radiacao for d in dd) / n, 1),
        "agua_solo_media_pct": round(sum(fracoes[i] for i in idx) / n * 100, 1),
        "capacidade_agua_solo_mm": float(capacidade_mm),
        "fertilidade_solo": float(fertilidade),
        "latitude": round(float(latitude), 4),
        "geadas_no_periodo": sum(1 for d in dd if d.tmin <= e["geada"]),
        "dias_encharcado": sum(1 for i in idx if encharcado[i]),
        "dias_floracao": floracao,
    }
